Fix pickle deserialization of ExperimentData

wshanks opened the pull request:
Previously, `ExperimentData` objects could be serialized and
deserialized using Python's `pickle` module, but deserialized objects
were not completely restored and an exception would be raised when doing
some operations like running analysis on the restored object.

nkanazawa1989 reviewed (commented):
Looks good to me. I'll approve once lint error is resolved.

wshanks commented:
@nkanazawa1989 I got sidetracked from this for a while.... The reason I had marked this as a draft was that I was not sure about the changes that I made to the test code.

The actual code fix is straightforward. I wanted to add a test though to verify the fix and check for future regressions. The test I added runs analysis on an experiment data object, then serializes/deserializes the experiment data through pickle, and runs analysis on the new experiment data object. I wanted to compare the two experiment data objects. I found that a simple comparison of the experiment data object contents would not give equality because results store their creation times as a field. Additionally, every result gets a result ID which is derived from a uuid which is also time-based. These result IDs are generated internally, so really there is no way to generate two experiment data objects with results that are identical.

To allow for my test to work, I added a `**kwargs` to `is_equivalent` which allows an extra option like `ignore_result_id` to be passed down to the different checkers. I was uncertain about a few things:

1. Does `ignore_result_id` need to be an option? Maybe for `is_equivalent` result IDs should just always be ignored? Is there a case where they should not be? I think you would need to be testing serialization/deserialization without additional processing for the IDs to stay the same. Maybe you could just use `__eq__` in that case though if you really cared about the IDs matching.
2. I pick out `ignore_result_id` in `_check_service_analysis_results` and `_check_result_table`. Maybe I should keep propagating it?
3. In `_check_result_table`, I turn the data frame into a list for comparison. This assumes that the results are in the same order for each table. Should I try to sort them?

I decided to always ignore the result IDs (making my second question irrelevant) and do sorting on the table entries. I am open to other suggestions.
